File: examples/python/split_pixel_payload_example.py
#!/usr/bin/env python3
"""Read or create a split PixelData payload with DicomSDL.

Usage:
    python split_pixel_payload_example.py
    python split_pixel_payload_example.py main-p10.dcm pixel-payload.bin --frame 0
    python split_pixel_payload_example.py --split source.dcm main-out.dcm pixel-payload.bin
    python split_pixel_payload_example.py --split source.dcm main-out.dcm pixel-payload.bin \
        --transfer-syntax RLELossless

The split-payload convention is private to DicomSDL runtime workflows:

* The main P10 DICOM stores (7FE0,0010) PixelData as a fixed 22-byte PIXDATA1
  placeholder metadata value.
* The sidecar payload stores the complete PixelData value bytes.
* For encapsulated transfer syntaxes, the sidecar starts with the Basic Offset
  Table item and includes all fragment items plus the sequence delimiter.

Use split_pixeldata_payload() for byte-preserving split. If transcoding is
needed, serialize to memory with write_bytes_with_transfer_syntax() first, then
split those bytes. After a viewer decodes the frames it needs, call
detach_pixeldata_payload() before releasing the caller-owned payload memory. Pass
keep_dump=True if you want the detached marker to retain PixelData dump text for
diagnostics.
"""

from __future__ import annotations

import argparse
import struct
import sys
from pathlib import Path
from typing import Sequence

import dicomsdl as dicom


def _pad_ui(text: str) -> bytes:
    raw = text.encode("ascii")
    if len(raw) % 2 == 1:
        raw += b"\x00"
    return raw


def _pack_explicit_le(
    group: int,
    element: int,
    vr: str,
    value: bytes,
) -> bytes:
    header = struct.pack("<HH", group, element) + vr.encode("ascii")
    if vr in {"OB", "OD", "OF", "OL", "OV", "OW", "SQ", "UC", "UR", "UT", "UN"}:
        return header + b"\x00\x00" + struct.pack("<I", len(value)) + value
    return header + struct.pack("<H", len(value)) + value


def _placeholder_value(vr: str, vl: int, payload_length: int) -> bytes:
    return (
        dicom.PIXELDATA_PAYLOAD_PLACEHOLDER_MAGIC
        + vr.encode("ascii")
        + struct.pack("<IQ", vl, payload_length)
    )


def _is_placeholder_metadata(value: bytes) -> bool:
    return (
        value.startswith(dicom.PIXELDATA_PAYLOAD_PLACEHOLDER_MAGIC)
        and len(value) == 22
    )


def _build_part10(transfer_syntax_uid: str, body: bytes) -> bytes:
    meta_ts = _pack_explicit_le(0x0002, 0x0010, "UI", _pad_ui(transfer_syntax_uid))
    meta_length = _pack_explicit_le(0x0002, 0x0000, "UL", struct.pack("<I", len(meta_ts)))
    return b"\x00" * 128 + b"DICM" + meta_length + meta_ts + body


def _build_demo_main_p10() -> bytes:
    body = b"".join(
        [
            _pack_explicit_le(0x0028, 0x0002, "US", struct.pack("<H", 1)),
            _pack_explicit_le(0x0028, 0x0004, "CS", b"MONOCHROME2 "),
            _pack_explicit_le(0x0028, 0x0008, "IS", b"1 "),
            _pack_explicit_le(0x0028, 0x0010, "US", struct.pack("<H", 1)),
            _pack_explicit_le(0x0028, 0x0011, "US", struct.pack("<H", 3)),
            _pack_explicit_le(0x0028, 0x0100, "US", struct.pack("<H", 16)),
            _pack_explicit_le(0x0028, 0x0101, "US", struct.pack("<H", 16)),
            _pack_explicit_le(0x0028, 0x0102, "US", struct.pack("<H", 15)),
            _pack_explicit_le(0x0028, 0x0103, "US", struct.pack("<H", 0)),
            _pack_explicit_le(
                0x7FE0, 0x0010, "OB", _placeholder_value("OW", 6, 6)
            ),
        ]
    )
    return _build_part10("1.2.840.10008.1.2.1", body)


def _hex_prefix(data: bytes, max_count: int = 16) -> str:
    prefix = " ".join(f"{byte:02x}" for byte in data[:max_count])
    return prefix + (" ..." if len(data) > max_count else "")


def _print_decode_summary(dicom_file: dicom.DicomFile, frame: int) -> None:
    pixel_data = dicom_file["PixelData"]
    if pixel_data.vr == dicom.VR.PX:
        try:
            encoded = dicom_file.encoded_pixel_frame_bytes(frame)
            suffix = f" [{_hex_prefix(encoded)}]" if encoded else ""
            print(f"Encoded frame {frame}: {len(encoded)} bytes{suffix}")
        except Exception as exc:
            print(f"Encoded frame {frame}: unavailable ({exc})")

    try:
        decoded = dicom_file.pixel_data(frame)
        suffix = f" [{_hex_prefix(decoded)}]" if decoded else ""
        print(f"Decoded frame {frame}: {len(decoded)} bytes{suffix}")
    except Exception as exc:
        print(f"Decoded frame {frame}: unavailable ({exc})")


def _print_attached_summary(dicom_file: dicom.DicomFile, frame: int) -> None:
    print(f"TransferSyntaxUID: {dicom_file.transfer_syntax_uid.value}")
    print(f"Rows x Columns: {dicom_file.Rows} x {dicom_file.Columns}")
    print(f"PixelData VR after attach: {dicom_file['PixelData'].vr.str()}")
    print(f"Attached payload: {dicom_file.has_attached_pixeldata_payload}")
    _print_decode_summary(dicom_file, frame)


def _run_read_split(
    name: str,
    main_p10: bytes | bytearray,
    payload: bytearray,
    frame: int,
    keep_dump: bool,
) -> int:
    dicom_file = dicom.read_bytes_with_pixeldata_payload(
        main_p10, payload, name=name, copy=False
    )

    print(f"Loaded split DICOM: {dicom_file.path}")
    _print_attached_summary(dicom_file, frame)

    dicom_file.detach_pixeldata_payload(keep_dump=keep_dump)
    payload.clear()

    print(f"Attached payload after detach: {dicom_file.has_attached_pixeldata_payload}")
    print(f"Rows still available after detach: {dicom_file.Rows}")
    if keep_dump:
        print("Detached marker keeps PixelData dump text for diagnostics.")
    print("After detach, do not call pixel decode APIs until a payload is attached again.")

    return 0


def _run_split_source(
    source_path: Path,
    main_out_path: Path,
    payload_out_path: Path,
    transfer_syntax: str | None,
    frame: int,
    keep_dump: bool,
) -> int:
    if transfer_syntax:
        source = dicom.read_file(source_path)
        transcoded = source.write_bytes_with_transfer_syntax(transfer_syntax)
        split = dicom.split_pixeldata_payload(
            [], transcoded, name=f"{source_path}:{transfer_syntax}"
        )
        print(f"Serialized to memory and split target transfer syntax: {transfer_syntax}")
    else:
        split = dicom.split_pixeldata_payload(
            [], source_path
        )
        print("Split source bytes with its current transfer syntax.")

    if not split.ok:
        raise RuntimeError(split.error_message)
    main_bytes = split.main_bytes
    payload_bytes = split.pixel_payload

    main_out_path.write_bytes(main_bytes)
    payload_out_path.write_bytes(payload_bytes)
    print(f"Wrote main DICOM: {main_out_path} ({len(main_bytes)} bytes)")
    print(f"Wrote PixelData payload: {payload_out_path} ({len(payload_bytes)} bytes)")

    placeholder_only = dicom.read_bytes(main_bytes, name="split-main-placeholder-check")
    if not _is_placeholder_metadata(placeholder_only["PixelData"].value_bytes()):
        raise RuntimeError(
            "split main DICOM does not contain the PIXDATA1 PixelData placeholder"
        )
    print("Verified main DICOM has the PIXDATA1 PixelData placeholder.")

    main_owner = bytearray(main_bytes)
    payload_owner = bytearray(payload_bytes)
    rejoined = dicom.read_bytes_with_pixeldata_payload(
        main_owner,
        payload_owner,
        name="split-roundtrip-check",
        copy=False,
    )
    print("Reattached split payload for a roundtrip check.")
    _print_attached_summary(rejoined, frame)

    rejoined.detach_pixeldata_payload(keep_dump=keep_dump)
    payload_owner.clear()
    print("Detached roundtrip payload; metadata remains available.")
    return 0


def main(argv: Sequence[str]) -> int:
    parser = argparse.ArgumentParser(
        description="Read or create a PIXDATA1 split PixelData payload",
        epilog=(
            "Without --split, provide main_p10 and pixel_payload to read an already "
            "split instance. With --split, provide source.dcm, main-out.dcm, and "
            "pixel-payload.bin to create split outputs. The optional "
            "--transfer-syntax value can be a UID or DicomSDL keyword such as "
            "ExplicitVRLittleEndian, ExplicitVRBigEndian, or RLELossless."
        ),
    )
    parser.add_argument(
        "--split",
        action="store_true",
        help="split a normal source DICOM into main DICOM and PixelData payload files",
    )
    parser.add_argument(
        "--transfer-syntax",
        help="target transfer syntax to use while splitting",
    )
    parser.add_argument(
        "--keep-dump",
        action="store_true",
        help="keep PixelData dump text in the detached marker",
    )
    parser.add_argument("--frame", type=int, default=0, help="Frame index to decode")
    parser.add_argument("paths", nargs="*", help="input/output paths; see usage examples")
    args = parser.parse_args(argv)

    if args.split:
        if len(args.paths) != 3:
            parser.error("--split requires: source.dcm main-out.dcm pixel-payload.bin")
        return _run_split_source(
            Path(args.paths[0]),
            Path(args.paths[1]),
            Path(args.paths[2]),
            args.transfer_syntax,
            args.frame,
            args.keep_dump,
        )

    if args.transfer_syntax:
        parser.error("--transfer-syntax is only valid with --split")

    if not args.paths:
        return _run_read_split(
            "built-in-split-pixel-payload-demo",
            bytearray(_build_demo_main_p10()),
            bytearray(b"\x34\x12\x56\x78\x9A\xBC"),
            args.frame,
            args.keep_dump,
        )

    if len(args.paths) != 2:
        parser.error("provide main_p10 and pixel_payload, or use --split")

    return _run_read_split(
        args.paths[0],
        bytearray(Path(args.paths[0]).read_bytes()),
        bytearray(Path(args.paths[1]).read_bytes()),
        args.frame,
        args.keep_dump,
    )


if __name__ == "__main__":
    raise SystemExit(main(sys.argv[1:]))

File: examples/split_pixel_payload_example.cpp
#include <algorithm>
#include <cstdint>
#include <cstdlib>
#include <exception>
#include <filesystem>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <iterator>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

#include <dicom.h>

using namespace dicom::literals;

/*
This example demonstrates the DicomSDL split PixelData convention.

There are two separate workflows:

1. Attach and read an already split instance:
   - main P10 DICOM contains (7FE0,0010) PixelData as a 22-byte
     "PIXDATA1" placeholder metadata value.
   - pixel-payload.bin contains the complete PixelData value bytes.
   - For encapsulated transfer syntaxes, that value starts with the Basic Offset
     Table item and includes all fragment items plus the sequence delimiter.

2. Split a normal DICOM instance:
   - split_pixeldata_payload() replaces the source PixelData element with the
     PIXDATA1 placeholder without reserializing the rest of the P10 bytes.
   - If you need transcoding, first serialize to memory with
     write_bytes_with_transfer_syntax(), then split those bytes.

The external pixel payload memory is caller-owned. Detach before freeing it.
detach_pixeldata_payload() keeps only the lightweight PIXDATA1 marker; pass true if you
want the detached marker to retain PixelData dump text for diagnostics.
*/

namespace {

void append_u16_le(std::vector<std::uint8_t>& out, std::uint16_t value) {
    out.push_back(static_cast<std::uint8_t>(value & 0xFFu));
    out.push_back(static_cast<std::uint8_t>((value >> 8) & 0xFFu));
}

void append_u32_le(std::vector<std::uint8_t>& out, std::uint32_t value) {
    out.push_back(static_cast<std::uint8_t>(value & 0xFFu));
    out.push_back(static_cast<std::uint8_t>((value >> 8) & 0xFFu));
    out.push_back(static_cast<std::uint8_t>((value >> 16) & 0xFFu));
    out.push_back(static_cast<std::uint8_t>((value >> 24) & 0xFFu));
}

void append_u64_le(std::vector<std::uint8_t>& out, std::uint64_t value) {
    for (int shift = 0; shift < 64; shift += 8) {
        out.push_back(static_cast<std::uint8_t>((value >> shift) & 0xFFu));
    }
}

void append_bytes(std::vector<std::uint8_t>& out,
    const std::vector<std::uint8_t>& bytes) {
    out.insert(out.end(), bytes.begin(), bytes.end());
}

std::vector<std::uint8_t> padded_ui(std::string uid) {
    if (uid.empty() || uid.back() != '\0') {
        uid.push_back('\0');
    }
    if ((uid.size() & 1u) != 0u) {
        uid.push_back('\0');
    }
    return std::vector<std::uint8_t>(uid.begin(), uid.end());
}

void append_explicit_vr_le_16(std::vector<std::uint8_t>& out,
    dicom::Tag tag, char vr0, char vr1,
    const std::vector<std::uint8_t>& value) {
    append_u16_le(out, tag.group());
    append_u16_le(out, tag.element());
    out.push_back(static_cast<std::uint8_t>(vr0));
    out.push_back(static_cast<std::uint8_t>(vr1));
    append_u16_le(out, static_cast<std::uint16_t>(value.size()));
    append_bytes(out, value);
}

void append_explicit_vr_le_32(std::vector<std::uint8_t>& out,
    dicom::Tag tag, char vr0, char vr1,
    const std::vector<std::uint8_t>& value) {
    append_u16_le(out, tag.group());
    append_u16_le(out, tag.element());
    out.push_back(static_cast<std::uint8_t>(vr0));
    out.push_back(static_cast<std::uint8_t>(vr1));
    append_u16_le(out, 0);
    append_u32_le(out, static_cast<std::uint32_t>(value.size()));
    append_bytes(out, value);
}

std::vector<std::uint8_t> build_part10(std::string transfer_syntax_uid,
    const std::vector<std::uint8_t>& body) {
    std::vector<std::uint8_t> meta_ts;
    append_explicit_vr_le_16(meta_ts, dicom::Tag(0x0002u, 0x0010u),
        'U', 'I', padded_ui(std::move(transfer_syntax_uid)));

    std::vector<std::uint8_t> meta_length_value;
    append_u32_le(meta_length_value, static_cast<std::uint32_t>(meta_ts.size()));

    std::vector<std::uint8_t> meta_length;
    append_explicit_vr_le_16(meta_length, dicom::Tag(0x0002u, 0x0000u),
        'U', 'L', meta_length_value);

    std::vector<std::uint8_t> out(128, 0);
    out.insert(out.end(), {'D', 'I', 'C', 'M'});
    append_bytes(out, meta_length);
    append_bytes(out, meta_ts);
    append_bytes(out, body);
    return out;
}

std::vector<std::uint8_t> placeholder_magic() {
    return std::vector<std::uint8_t>(
        dicom::kPixelDataPayloadPlaceholderMagic.begin(),
        dicom::kPixelDataPayloadPlaceholderMagic.end());
}

std::vector<std::uint8_t> placeholder_value(
    char vr0, char vr1, std::uint32_t vl, std::uint64_t payload_length) {
    auto value = placeholder_magic();
    value.push_back(static_cast<std::uint8_t>(vr0));
    value.push_back(static_cast<std::uint8_t>(vr1));
    append_u32_le(value, vl);
    append_u64_le(value, payload_length);
    return value;
}

std::vector<std::uint8_t> build_demo_main_p10() {
    std::vector<std::uint8_t> body;
    append_explicit_vr_le_16(body, "SamplesPerPixel"_tag, 'U', 'S',
        {0x01u, 0x00u});
    append_explicit_vr_le_16(body, "PhotometricInterpretation"_tag, 'C', 'S',
        {'M', 'O', 'N', 'O', 'C', 'H', 'R', 'O', 'M', 'E', '2', ' '});
    append_explicit_vr_le_16(body, "NumberOfFrames"_tag, 'I', 'S', {'1', ' '});
    append_explicit_vr_le_16(body, "Rows"_tag, 'U', 'S', {0x01u, 0x00u});
    append_explicit_vr_le_16(body, "Columns"_tag, 'U', 'S', {0x03u, 0x00u});
    append_explicit_vr_le_16(body, "BitsAllocated"_tag, 'U', 'S',
        {0x10u, 0x00u});
    append_explicit_vr_le_16(body, "BitsStored"_tag, 'U', 'S',
        {0x10u, 0x00u});
    append_explicit_vr_le_16(body, "HighBit"_tag, 'U', 'S', {0x0Fu, 0x00u});
    append_explicit_vr_le_16(body, "PixelRepresentation"_tag, 'U', 'S',
        {0x00u, 0x00u});

    // The main P10 keeps only placeholder metadata in PixelData.
    append_explicit_vr_le_32(body, "PixelData"_tag, 'O', 'B',
        placeholder_value('O', 'W', 6u, 6u));

    return build_part10("1.2.840.10008.1.2.1", body);
}

std::vector<std::uint8_t> read_all_bytes(const std::string& path) {
    std::ifstream in(path, std::ios::binary);
    if (!in) {
        throw std::runtime_error("failed to open: " + path);
    }
    return std::vector<std::uint8_t>(
        std::istreambuf_iterator<char>(in), std::istreambuf_iterator<char>());
}

void write_all_bytes(const std::string& path,
    const std::vector<std::uint8_t>& bytes) {
    std::ofstream out(path, std::ios::binary);
    if (!out) {
        throw std::runtime_error("failed to create: " + path);
    }
    if (!bytes.empty()) {
        out.write(reinterpret_cast<const char*>(bytes.data()),
            static_cast<std::streamsize>(bytes.size()));
    }
    if (!out) {
        throw std::runtime_error("failed to write: " + path);
    }
}

template <typename ByteRange>
std::string hex_prefix(const ByteRange& bytes, std::size_t max_count = 16) {
    std::ostringstream out;
    const auto count = std::min(bytes.size(), max_count);
    for (std::size_t i = 0; i < count; ++i) {
        if (i != 0) {
            out << ' ';
        }
        out << std::hex << std::setw(2) << std::setfill('0')
            << static_cast<unsigned>(bytes[i]);
    }
    if (bytes.size() > count) {
        out << " ...";
    }
    return out.str();
}

bool is_placeholder_pixel_data(const dicom::DataElement& pixel_data) {
    const auto value = pixel_data.value_span();
    return value.size() == dicom::kPixelDataPayloadPlaceholderMetadataSize &&
        std::equal(dicom::kPixelDataPayloadPlaceholderMagic.begin(),
            dicom::kPixelDataPayloadPlaceholderMagic.end(), value.begin());
}

void print_decode_summary(dicom::DicomFile& file, std::size_t frame_index) {
    try {
        const auto decoded = file.pixel_data(frame_index);
        std::cout << "Decoded frame " << frame_index << ": "
                  << decoded.size() << " bytes";
        if (!decoded.empty()) {
            std::cout << " [" << hex_prefix(decoded) << "]";
        }
        std::cout << "\n";
    } catch (const std::exception& ex) {
        std::cout << "Decoded frame " << frame_index
                  << ": unavailable (" << ex.what() << ")\n";
    }
}

void print_encoded_summary_if_available(dicom::DicomFile& file,
    std::size_t frame_index) {
    auto& pixel_data = file.get_dataelement("PixelData"_tag);
    if (!pixel_data.vr().is_pixel_sequence()) {
        return;
    }
    try {
        const auto encoded = file.encoded_pixel_frame_view(frame_index);
        std::cout << "Encoded frame " << frame_index << ": "
                  << encoded.size() << " bytes";
        if (!encoded.empty()) {
            std::cout << " [" << hex_prefix(encoded) << "]";
        }
        std::cout << "\n";
    } catch (const std::exception& ex) {
        std::cout << "Encoded frame " << frame_index
                  << ": unavailable (" << ex.what() << ")\n";
    }
}

void print_attached_summary(dicom::DicomFile& file, std::size_t frame_index) {
    auto& dataset = file.dataset();
    auto& pixel_data = file.get_dataelement("PixelData"_tag);

    std::cout << "TransferSyntaxUID: "
              << (file.transfer_syntax_uid().valid()
                     ? file.transfer_syntax_uid().value()
                     : "<missing>")
              << "\n";
    std::cout << "Rows x Columns: "
              << dataset["Rows"_tag].to_long().value_or(0) << " x "
              << dataset["Columns"_tag].to_long().value_or(0) << "\n";
    std::cout << "PixelData VR after attach: " << pixel_data.vr().str()
              << "\n";
    std::cout << "Attached payload: "
              << (file.has_attached_pixeldata_payload() ? "yes" : "no") << "\n";

    print_encoded_summary_if_available(file, frame_index);
    print_decode_summary(file, frame_index);
}

int run_read_split_example(const std::string& name,
    std::vector<std::uint8_t>& main_p10,
    std::vector<std::uint8_t>& pixel_payload, std::size_t frame_index) {
    auto file = dicom::read_bytes_with_pixeldata_payload(name,
        main_p10.data(), main_p10.size(),
        pixel_payload.data(), pixel_payload.size());

    std::cout << "Loaded split DICOM: " << file->path() << "\n";
    print_attached_summary(*file, frame_index);

    // After decode, the caller can detach and release the external payload.
    // Pass true to keep PixelData dump text inside the detached marker.
    file->detach_pixeldata_payload();
    pixel_payload.clear();
    pixel_payload.shrink_to_fit();

    std::cout << "Attached payload after detach: "
              << (file->has_attached_pixeldata_payload() ? "yes" : "no") << "\n";
    std::cout << "Rows still available after detach: "
              << file->dataset()["Rows"_tag].to_long().value_or(0) << "\n";
    std::cout << "After detach, do not call pixel decode APIs until a payload "
              << "is attached again.\n";

    return 0;
}

int run_split_source_example(const std::string& source_path,
    const std::string& main_out_path, const std::string& payload_out_path,
    const std::string& transfer_syntax_text, std::size_t frame_index) {
    dicom::SplitPixelDataPayloadResult split;

    if (transfer_syntax_text.empty()) {
        split = dicom::split_pixeldata_payload(
            dicom::DataSetSelection{}, std::filesystem::path(source_path));
        std::cout << "Split source bytes with its current transfer syntax.\n";
    } else {
        auto source = dicom::read_file(source_path);
        const auto transfer_syntax = dicom::uid::lookup_or_throw(transfer_syntax_text);
        const auto transcoded =
            source->write_bytes_with_transfer_syntax(transfer_syntax);
        split = dicom::split_pixeldata_payload(
            dicom::DataSetSelection{}, "transcoded-split-source",
            std::span<const std::uint8_t>(
                transcoded.data(), transcoded.size()));
        std::cout << "Serialized to memory and split target transfer syntax: "
                  << transfer_syntax.value() << "\n";
    }
    if (!split.ok()) {
        throw std::runtime_error(split.error_message.empty()
                ? std::string("split_pixeldata_payload failed")
                : split.error_message);
    }

    write_all_bytes(main_out_path, split.main_bytes);
    write_all_bytes(payload_out_path, split.pixel_payload);

    std::cout << "Wrote main DICOM: " << main_out_path << " ("
              << split.main_bytes.size() << " bytes)\n";
    std::cout << "Wrote PixelData payload: " << payload_out_path << " ("
              << split.pixel_payload.size() << " bytes)\n";

    auto placeholder_only = dicom::read_bytes("split-main-placeholder-check",
        split.main_bytes.data(), split.main_bytes.size());
    if (!is_placeholder_pixel_data(
            placeholder_only->get_dataelement("PixelData"_tag))) {
        throw std::runtime_error(
            "split main DICOM does not contain the PIXDATA1 PixelData placeholder");
    }
    std::cout << "Verified main DICOM has the PIXDATA1 PixelData placeholder.\n";

    auto rejoined = dicom::read_bytes_with_pixeldata_payload(
        "split-roundtrip-check", split.main_bytes.data(),
        split.main_bytes.size(), split.pixel_payload.data(),
        split.pixel_payload.size());
    std::cout << "Reattached split payload for a roundtrip check.\n";
    print_attached_summary(*rejoined, frame_index);

    rejoined->detach_pixeldata_payload();
    std::cout << "Detached roundtrip payload; metadata remains available.\n";
    return 0;
}

void print_usage(const char* program) {
    std::cerr
        << "Usage:\n"
        << "  " << program << "\n"
        << "  " << program << " <main-p10.dcm> <pixel-payload.bin> [frame]\n"
        << "  " << program << " --split <source.dcm> <main-out.dcm>"
        << " <pixel-payload-out.bin> [frame]\n"
        << "  " << program << " --split-ts <source.dcm> <transfer-syntax>"
        << " <main-out.dcm> <pixel-payload-out.bin> [frame]\n\n"
        << "No arguments:\n"
        << "  Run a tiny built-in native PixelData attach/detach demo.\n\n"
        << "Read split inputs:\n"
        << "  main-p10.dcm must contain (7FE0,0010) PixelData as the fixed\n"
        << "  22-byte PIXDATA1 placeholder metadata. pixel-payload.bin must contain the\n"
        << "  complete PixelData value bytes. For encapsulated payloads, that\n"
        << "  means Basic Offset Table item, fragment items, and sequence\n"
        << "  delimiter.\n\n"
        << "Create split outputs:\n"
        << "  --split keeps the source transfer syntax. --split-ts serializes\n"
        << "  through a target transfer syntax before splitting. Use a UID value\n"
        << "  or keyword such as ExplicitVRLittleEndian, ExplicitVRBigEndian,\n"
        << "  or RLELossless.\n";
}

} // namespace

int main(int argc, char* argv[]) {
    try {
        if (argc == 1) {
            auto main_p10 = build_demo_main_p10();
            std::vector<std::uint8_t> pixel_payload{
                0x34u, 0x12u, 0x56u, 0x78u, 0x9Au, 0xBCu};
            return run_read_split_example(
                "built-in-split-pixel-payload-demo", main_p10, pixel_payload, 0);
        }

        const std::string command = argv[1];
        if (command == "--split") {
            if (argc < 5 || argc > 6) {
                print_usage(argv[0]);
                return 1;
            }
            const auto frame_index = argc == 6
                ? static_cast<std::size_t>(std::stoull(argv[5]))
                : std::size_t{0};
            return run_split_source_example(
                argv[2], argv[3], argv[4], {}, frame_index);
        }

        if (command == "--split-ts") {
            if (argc < 6 || argc > 7) {
                print_usage(argv[0]);
                return 1;
            }
            const auto frame_index = argc == 7
                ? static_cast<std::size_t>(std::stoull(argv[6]))
                : std::size_t{0};
            return run_split_source_example(
                argv[2], argv[4], argv[5], argv[3], frame_index);
        }

        if (argc < 3 || argc > 4) {
            print_usage(argv[0]);
            return 1;
        }

        auto main_p10 = read_all_bytes(argv[1]);
        auto pixel_payload = read_all_bytes(argv[2]);
        const auto frame_index = argc == 4
            ? static_cast<std::size_t>(std::stoull(argv[3]))
            : std::size_t{0};

        return run_read_split_example(argv[1], main_p10, pixel_payload, frame_index);
    } catch (const std::exception& ex) {
        std::cerr << "split_pixel_payload_example: " << ex.what() << "\n";
        return 1;
    }
}
